File: app/src/test/java/com/caglar/pokequery/audit/PreAabAuditRegressionTest.kt
package com.caglar.pokequery.audit

import com.caglar.pokequery.domain.assist.SearchIntentParser
import com.caglar.pokequery.domain.assist.SearchStringExplainer
import com.caglar.pokequery.domain.engine.GoalStringBuilder
import com.caglar.pokequery.domain.engine.StringBuilderEngine
import com.caglar.pokequery.domain.lint.Linter
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

class PreAabAuditRegressionTest {

    @Test
    fun `negation is scoped to the matched clause`() {
        val parsed = SearchIntentParser.parse("Find hundos and exclude shinies")
        assertEquals("4*&!shiny", parsed.rawQuery)
        assertTrue("4*" in parsed.tokens)
        assertTrue("shiny" in parsed.exclusions)
        assertFalse("4*" in parsed.exclusions)
    }

    @Test
    fun `localized goal protections are not duplicated`() {
        val base = StringBuilderEngine.buildGoal("safe_cleanup", language = "Turkish")
        val merged = GoalStringBuilder.buildFinal(
            base,
            optionalProtections = listOf("shiny", "legendary", "traded"),
            language = "Turkish"
        )
        val tokens = merged.rawSyntax.split('&')
        assertEquals(1, tokens.count { it == "!parlak" })
        assertEquals(1, tokens.count { it == "!efsanevi" })
        assertEquals(1, tokens.count { it == "!takas edilen" })
    }

    @Test
    fun `count protection checks exact tokens not substrings`() {
        val otherProtections = StringBuilderEngine.COUNT_MANDATORY_PROTECTIONS
            .filterNot { it == "shiny" }
            .joinToString("&") { "!$it" }
        val warnings = Linter.lint("count2-&!shinyx&$otherProtections")
        assertTrue(warnings.any { it.isError && "!shiny" in it.message })
    }

    @Test
    fun `unsupported untraded token is blocked`() {
        val warnings = Linter.lint("!untraded")
        assertTrue(warnings.any { it.isError && "untraded" in it.message })
    }

    @Test
    fun `standard range tokens are explained as known structured filters`() {
        val explained = SearchStringExplainer.explain("count2-&0-1attack&3-4defense&3-4hp")
        assertFalse(explained.hasUnknownTokens)
        assertTrue(explained.tokens.any { it.token == "count2-" && it.category == "count_filter" })
        assertTrue(explained.tokens.any { it.token == "0-1attack" && it.category == "iv_stat" })
        assertTrue(explained.tokens.any { it.token == "3-4defense" && it.category == "iv_stat" })
        assertTrue(explained.tokens.any { it.token == "3-4hp" && it.category == "iv_stat" })
    }
}

File: web/src/engine/preAabAudit.regression.test.ts
import { describe, expect, it } from 'vitest'
import { parseSearchIntent } from './searchIntentParser'
import { buildFinal } from './goalStringBuilder'
import { buildGoal, COUNT_MANDATORY_PROTECTIONS } from './stringBuilderEngine'
import { lint } from './linter'
import { explain } from './searchStringExplainer'

describe('pre-AAB audit regressions', () => {
  it('scopes negation to the matched clause', () => {
    const parsed = parseSearchIntent('Find hundos and exclude shinies')
    expect(parsed.rawQuery).toBe('4*&!shiny')
    expect(parsed.tokens).toContain('4*')
    expect(parsed.exclusions).toContain('shiny')
    expect(parsed.exclusions).not.toContain('4*')
  })

  it('does not duplicate localized protections', () => {
    const base = buildGoal('safe_cleanup', '', '', 'Turkish')
    const merged = buildFinal(base, ['shiny', 'legendary', 'traded'], 'Turkish')
    const tokens = merged.rawSyntax.split('&')
    expect(tokens.filter(t => t === '!parlak')).toHaveLength(1)
    expect(tokens.filter(t => t === '!efsanevi')).toHaveLength(1)
    expect(tokens.filter(t => t === '!takas edilen')).toHaveLength(1)
  })

  it('does not accept substring-spoofed count protections', () => {
    const other = COUNT_MANDATORY_PROTECTIONS
      .filter(p => p !== 'shiny')
      .map(p => `!${p}`)
      .join('&')
    const warnings = lint(`count2-&!shinyx&${other}`)
    expect(warnings.some(w => w.isError && w.message.includes('!shiny'))).toBe(true)
  })

  it('blocks the unsupported untraded token', () => {
    const warnings = lint('!untraded')
    expect(warnings.some(w => w.isError && w.message.includes('untraded'))).toBe(true)
  })

  it('recognizes normal count and IV range tokens', () => {
    const result = explain('count2-&0-1attack&3-4defense&3-4hp')
    expect(result.hasUnknownTokens).toBe(false)
    expect(result.tokens).toEqual(expect.arrayContaining([
      expect.objectContaining({ token: 'count2-', category: 'count_filter' }),
      expect.objectContaining({ token: '0-1attack', category: 'iv_stat' }),
      expect.objectContaining({ token: '3-4defense', category: 'iv_stat' }),
      expect.objectContaining({ token: '3-4hp', category: 'iv_stat' }),
    ]))
  })
})
